### DISCUSSION TITLE
Guide to generate machine independent initrds?

### DISCUSSION THREAD

**Discussion by septatrix** [systemd/mkosi#2403] [Q&A] [answered]
Hey everyone, based on several talk and blogs it seems that some distributions want to move into the direction where they generate initrd for all users and then distribute them. This has some advantages such as vendor-signed UKIs and a more consistent setup which might help resolving bugs.

However, I have not been able to find information how such initrds might be generated due to the vast amount of different hardware which has to be supported. Which modules are actually required to be included in an initrd? There is for example the list in this repositories mkosi-initrd configuration:

https://github.com/systemd/mkosi/blob/ae8e9053d2bb34214fb114857875b90831c15c8f/mkosi/resources/mkosi-initrd/mkosi.conf#L39-L50

Is most hardware fine if it is only brought up after the initrd when the real system starts? Some other projects go another way and simply include based on a very coarse list (but also not listing any drivers):

https://github.com/swsnr/arch-rescue-image/blob/40af4559b7f792e6d6ca43c4cd6fe894f3048f68/mkosi.conf#L58-L65

Or Ubuntu's list for modules which are included in their initrds distributed via snap for Ubuntu Core which seems like a very carefully curated list of modules:

https://github.com/snapcore/core-initrd/blob/95ade526c631fed37b83e676560057bcc488620d/modules/main/extra-modules.conf

Is there any guide upon this topic which you know? Does one simply start with the minimal set and add more if bug reports start pouring in?

**Comment by daandemeyer** [comment, marked as answer]
There is no magic solution here but trial and error to figure out what should be in and what shouldn't be. Not sure if there's more to say about the topic.

**Comment by septatrix** [comment]
In addition to the list from the snapcore initrd I also found out that initramfs tools has a set of kernle modules it installs when selecting `MODULES=most` (as opposed to "dep", which seems to be similar hostonly mode):

[initramfs-tools/hook-functions](https://salsa.debian.org/kernel-team/initramfs-tools/-/blob/cf964bfb4362019fd7fba1e839e403ff950dca8e/hook-functions#L595)

Dracut on the other hand seems to scan all available modules and then determines (have not found out how exactly) those used for block devices, filesystems, keyboard drivers etc. and installs those:

https://wiki.gentoo.org/wiki/Dracut#Kernel_modules

PS:

The dracut logic seems to be in the kernel-modules dracut module: [90kernel-modules/module-setup.sh](https://github.com/dracutdevs/dracut/blob/4980bad34775da715a2639b736cba5e65a8a2604/modules.d/90kernel-modules/module-setup.sh)

**Reply by septatrix** [reply]
Another resource on this is Fedora's [Dracut configuration for their kernel-virt-uki](https://src.fedoraproject.org/rpms/kernel/blob/b9352d952c6b6ac4e9b7411d7b4480a91994b193/f/dracut-virt.conf). It is obviously not suitable for general purposes hardware but ought to be sufficient for VMs.
